sqlmesh/core/schema_diff.py

```diff
@@ -14,83 +17,366 @@ class SchemaDeltaOp(Enum):
-        """
-        to_schema = self.engine_adapter.columns(apply_to_table)
-        from_schema = self.engine_adapter.columns(schema_from_table)
-
-        result = []
-
-        for to_column_name, to_column_type in to_schema.items():
-            from_column_type = from_schema.get(to_column_name)
-            from_column_type = from_column_type.upper() if from_column_type else None
-            if from_column_type != to_column_type.upper():
-                if from_column_type and self.is_type_transition_allowed(
-                    to_column_type, from_column_type
-                ):
-                    result.append(SchemaDelta.alter_type(to_column_name, from_column_type))
-                else:
-                    result.append(SchemaDelta.drop(to_column_name, to_column_type))
-                    if from_column_type:
-                        result.append(SchemaDelta.add(to_column_name, from_column_type))
-
-        for from_column_name, from_column_type in from_schema.items():
-            if from_column_name not in to_schema:
-                result.append(SchemaDelta.add(from_column_name, from_column_type))
-
-        return result
+    def get_matching_kwarg(
+        current_kwarg: exp.StructKwarg, new_struct: exp.DataType, current_pos: int
+    ) -> t.Tuple[t.Optional[int], t.Optional[exp.StructKwarg]]:
+        current_name, current_type = _get_name_and_type(current_kwarg)
+        # First check if we have the same column in the same position to get O(1) complexity
+        new_kwarg = seq_get(new_struct.expressions, current_pos)
+        if new_kwarg:
+            new_name, new_type = _get_name_and_type(new_kwarg)
+            if current_name == new_name:
+                return current_pos, new_kwarg
+        # If not, check if we have the same column in a all positions to get O(n) complexity
+        for i, new_kwarg in enumerate(new_struct.expressions):
+            new_name, new_type = _get_name_and_type(new_kwarg)
+            if current_name == new_name:
+                return i, new_kwarg
+        return None, None
+
+    def get_column_name(name: str, parents: ParentColumns) -> str:
+        if parents.has_columns:
+            return ".".join([parents.sql(), name])
+        return name
+
+    def get_alter_op(
+        pos: int,
+        struct: exp.DataType,
+        parent_columns: ParentColumns,
+        name: str,
+        new_type: exp.DataType,
+        current_type: t.Union[str, exp.DataType],
+    ) -> SchemaDelta:
+        col_pos = ColumnPosition.create(pos, struct.expressions, replacing_col=True)
+        return SchemaDelta.alter_type(
+            get_column_name(name, parent_columns), new_type, current_type, col_pos, parent_columns
+        )
+
+    current_struct = current_struct.copy()
+    parent_columns = parent_columns or ParentColumns(columns=[])
+    operations = []
+    # Resolve all drop columns
+    pop_offset = 0
+    for current_pos, current_kwarg in enumerate(current_struct.expressions[:]):
```

Comment:
what does [:] do?

Reply:
Makes it so I'm iterating over a copy since the original is being modified while looping

Reply:
can you just call copy() on it, a bit more clear